File: Java Project  V1.7/backend-spring/src/main/java/com/emart/service/purchase/CartDecision.java
package com.emart.service.purchase;

import java.math.BigDecimal;
import java.util.List;

/**
 * The decision for a WHOLE cart: every line decided independently
 * (they may each be a different purchase mode), then aggregated.
 *
 * Immutable. Cart responses, the checkout guard and the invoice totals
 * all read this one object, which is what keeps "what the cart showed"
 * and "what checkout charged" identical.
 */
public final class CartDecision {

    private final List<LineDecision> lines;

    /** Sum of regular (pre-offer) line totals - the MRP subtotal. */
    private final BigDecimal subtotal;

    /** Cash actually payable across the cart. */
    private final BigDecimal payableTotal;

    /** Points redeemed across the cart. */
    private final int totalPointsRequired;

    /** Points earned on the cash actually paid. */
    private final int pointsEarned;

    private final int openingBalance;

    private final boolean purchasable;
    private final String blockingReason;


    public CartDecision(
            List<LineDecision> lines,
            BigDecimal subtotal,
            BigDecimal payableTotal,
            int totalPointsRequired,
            int pointsEarned,
            int openingBalance,
            boolean purchasable,
            String blockingReason) {

        this.lines = lines == null ? List.of() : List.copyOf(lines);
        this.subtotal = nvl(subtotal);
        this.payableTotal = nvl(payableTotal);
        this.totalPointsRequired = Math.max(0, totalPointsRequired);
        this.pointsEarned = Math.max(0, pointsEarned);
        this.openingBalance = Math.max(0, openingBalance);
        this.purchasable = purchasable;
        this.blockingReason = blockingReason;
    }


    public List<LineDecision> getLines() {
        return lines;
    }

    public BigDecimal getSubtotal() {
        return subtotal;
    }

    public BigDecimal getPayableTotal() {
        return payableTotal;
    }

    public int getTotalPointsRequired() {
        return totalPointsRequired;
    }

    public int getPointsEarned() {
        return pointsEarned;
    }

    public int getOpeningBalance() {
        return openingBalance;
    }

    public boolean isPurchasable() {
        return purchasable;
    }

    public String getBlockingReason() {
        return blockingReason;
    }


    /** subtotal - payableTotal, i.e. the cash saved across the cart. */
    public BigDecimal getTotalSavings() {
        return subtotal.subtract(payableTotal);
    }

    /** opening - redeemed + earned. Never negative by construction. */
    public int getClosingBalance() {
        return Math.max(0, openingBalance - totalPointsRequired + pointsEarned);
    }


    private static BigDecimal nvl(BigDecimal value) {
        return value == null ? BigDecimal.ZERO : value;
    }
}

File: DotNet Project V2.0/backend-dotnet/src/Emart.Domain/Purchase/CartDecision.cs
namespace Emart.Domain.Purchase;

/// <summary>
/// The decision for a WHOLE cart: every line decided independently (they may each be a different
/// purchase mode), then aggregated. Immutable.
/// </summary>
public sealed class CartDecision
{
    public IReadOnlyList<LineDecision> Lines { get; }

    /// <summary>Sum of regular (pre-offer) line totals - the MRP subtotal.</summary>
    public decimal Subtotal { get; }

    /// <summary>Cash actually payable across the cart.</summary>
    public decimal PayableTotal { get; }

    /// <summary>Points redeemed across the cart.</summary>
    public int TotalPointsRequired { get; }

    /// <summary>Points earned on the cash actually paid.</summary>
    public int PointsEarned { get; }

    public int OpeningBalance { get; }

    public bool Purchasable { get; }
    public string? BlockingReason { get; }

    public CartDecision(
        IReadOnlyList<LineDecision>? lines,
        decimal subtotal,
        decimal payableTotal,
        int totalPointsRequired,
        int pointsEarned,
        int openingBalance,
        bool purchasable,
        string? blockingReason)
    {
        Lines = lines ?? Array.Empty<LineDecision>();
        Subtotal = subtotal;
        PayableTotal = payableTotal;
        TotalPointsRequired = Math.Max(0, totalPointsRequired);
        PointsEarned = Math.Max(0, pointsEarned);
        OpeningBalance = Math.Max(0, openingBalance);
        Purchasable = purchasable;
        BlockingReason = blockingReason;
    }

    /// <summary>subtotal - payableTotal, i.e. the cash saved across the cart.</summary>
    public decimal TotalSavings => Subtotal - PayableTotal;

    /// <summary>opening - redeemed + earned. Never negative by construction.</summary>
    public int ClosingBalance => Math.Max(0, OpeningBalance - TotalPointsRequired + PointsEarned);
}
